cloudinit/sources/DataSourceEc2.py

```diff
@@ -891,30 +894,35 @@ def _collect_platform_data():
 
 
 def _build_nic_order(
-    macs_metadata: Dict[str, Dict], macs: List[str]
+    macs_metadata: Dict[str, Dict],
+    macs_to_nics: Dict[str, str],
+    default_nic_order: NicOrder = NicOrder.MAC,
 ) -> Dict[str, int]:
     """
-    Builds a dictionary containing macs as keys nad nic orders as values,
+    Builds a dictionary containing macs as keys and nic orders as values,
     taking into account `network-card` and `device-number` if present.
 
     Note that the first NIC will be the primary NIC as it will be the one with
     [network-card] == 0 and device-number == 0 if present.
 
     @param macs_metadata: dictionary with mac address as key and contents like:
     {"device-number": "0", "interface-id": "...", "local-ipv4s": ...}
-    @macs: list of macs to consider
+    @macs_to_nics: dictionary with mac address as key and nic name as value
 
     @return: Dictionary with macs as keys and nic orders as values.
     """
     nic_order: Dict[str, int] = {}
+    macs = sorted(macs_to_nics.keys())
```

Comment:
Sorting the macs is not required, as we sort by mac or nic_name later on, right? So, let's simplify it:

```diff
diff --git a/cloudinit/sources/DataSourceEc2.py b/cloudinit/sources/DataSourceEc2.py
index e9458713f..64c328737 100644
--- a/cloudinit/sources/DataSourceEc2.py
+++ b/cloudinit/sources/DataSourceEc2.py
@@ -914,8 +914,7 @@ def _build_nic_order(
     @return: Dictionary with macs as keys and nic orders as values.
     """
     nic_order: Dict[str, int] = {}
-    macs = sorted(macs_to_nics.keys())
-    if len(macs) == 0 or len(macs_metadata) == 0:
+    if len(macs_to_nics) == 0 or len(macs_metadata) == 0:
         return nic_order
 
     valid_macs_metadata = filter(
@@ -923,7 +922,8 @@ def _build_nic_order(
         lambda mmd: mmd[1] is not None,
         # filter by macs
         map(
-            lambda mac: (mac, macs_metadata.get(mac), macs_to_nics[mac]), macs
+            lambda mac: (mac, macs_metadata.get(mac), macs_to_nics[mac]),
+            macs_to_nics.keys(),
         ),
     )
 
@@ -941,7 +941,7 @@ def _build_nic_order(
     # function.
     return {
         mac: i
-        for i, (mac, _mac_metadata, nic_name) in enumerate(
+        for i, (mac, _mac_metadata, _nic_name) in enumerate(
             sorted(
                 valid_macs_metadata,
                 key=lambda mmd: (
```

Reply:
I have already change it. thanks